src/query/src/range_select/plan.rs

```diff
@@ -117,32 +120,88 @@ impl Fill {
-                                        continue;
-                                    }
-                                }
-                            }
-                        }
-                    }
+                    Fill::Linear => {}
                     Fill::Const(v) => data[i] = v.clone(),
                 }
             }
         }
         Ok(())
     }
+
+    fn fill_linear(ts: &[i64], data: &mut [ScalarValue]) -> DfResult<()> {
+        let not_null_num = data
+            .iter()
+            .fold(0, |acc, x| if x.is_null() { acc } else { acc + 1 });
+        // We need at least two non-empty data points to perform linear interpolation
+        if not_null_num < 2 {
+            return Ok(());
+        }
+        let mut index = 0;
+        while index < data.len() {
+            // find null interval [start, end)
+            let start = data[index..]
+                .iter()
+                .position(ScalarValue::is_null)
+                .unwrap_or(data.len() - index)
+                + index;
+            if start == data.len() {
+                break;
+            }
+            let end = data[start..]
+                .iter()
+                .position(|r| !r.is_null())
+                .unwrap_or(data.len() - start)
+                + start;
+            index = end + 1;
+            // head or tail null dispose later
+            if start != 0 && end != data.len() {
+                linear_interpolation(ts, data, start - 1, end, start, end)?;
+            }
+        }
+        // dispose head null interval
+        if data[0].is_null() {
+            let start = data.iter().position(|r| !r.is_null()).unwrap();
+            linear_interpolation(ts, data, start, start + 1, 0, start)?;
+        }
+        // dispose tail null interval
+        if data[data.len() - 1].is_null() {
+            let start = data.len() - 1 - data.iter().rev().position(|r| !r.is_null()).unwrap();
+            linear_interpolation(ts, data, start, start - 1, start + 1, data.len())?;
+        }
+        Ok(())
+    }
+}
+
+/// use `(ts[i1], data[i1])`, `(ts[i2], data[i2])` as endpoint, linearly interpolates element over the interval `[start, end)`
+fn linear_interpolation(
+    ts: &[i64],
+    data: &mut [ScalarValue],
+    i1: usize,
+    i2: usize,
+    start: usize,
+    end: usize,
+) -> DfResult<()> {
+    let (x0, x1) = (ts[i1] as f64, ts[i2] as f64);
+    let (y0, y1, is_float32) = match (&data[i1], &data[i2]) {
+        (ScalarValue::Float64(Some(y0)), ScalarValue::Float64(Some(y1))) => (*y0, *y1, false),
+        (ScalarValue::Float32(Some(y0)), ScalarValue::Float32(Some(y1))) => {
+            (*y0 as f64, *y1 as f64, true)
+        }
+        _ => {
+            return Err(DataFusionError::Execution(
+                "RangePlan: Apply Fill LINEAR strategy on Non-floating type".to_string(),
+            ));
+        }
+    };
+    for i in start..end {
+        let val = y0 + (y1 - y0) / (x1 - x0) * (ts[i] as f64 - x0);
```

Comment:
May the dividend be zero? We can do defensive programming here.

Reply:
It can't happen in theory, but we can do that